src/renderers/shared/fiber/ReactFiberBeginWork.js

```diff
@@ -1154,7 +1165,29 @@ module.exports = function<T, P, I, TI, PI, C, CX, PL>(
   ): void {
     const progressedPriority = workInProgress.progressedPriority;
     const progressedWork = workInProgress.progressedWork;
-    if (progressedPriority === renderPriority && progressedWork !== current) {
+    const newestWork = workInProgress.newestWork;
+
+    if (
+      // Only resume on the progressed work if it rendered at the
+      // same priority.
+      progressedPriority === renderPriority &&
+      // Don't resume on current; use the reset path below.
+      progressedWork !== current &&
+      // It's possible for a work-in-progress fiber to be the most progressed
+      // work (last fiber whose children were reconciled) but be older than
+      // the current commit. This scenario, where the work-in-progress fiber
+      // is really the "previous current," happens after a bailout. To avoid,
+      // in addition to keeping track of the most progressed fiber, we also
+      // keep track of the most recent fiber to enter the begin phase,
+      // regardless of whether it bailed out. Then we only resume on the work-
+      // in-progress if its the most recent fiber.
+      // TODO: It'd be nice to come up with a heuristic here that didn't require
+      // an additional fiber field.
+      !(progressedWork === workInProgress && workInProgress !== newestWork)
```

Comment:
Interesting, I wrote it that way at first but then rewrote it this way because I thought it was easier to understand haha :D